### PR TITLE
Redis client

### PR DESCRIPTION
**Pull request by 115100**

### CODE REVIEW

File: client/manager_redis.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+	"github.com/pkg/errors"
+	"gopkg.in/redis.v5"
+)
+
+type RedisManager struct {
+	DB     *redis.Client
+	Hasher fosite.Hasher
+}
+
+const redisClientTemplate = "hydra:client:%s"
```
Comment (dim) [inline client/manager_redis.go]: Use `"hydra:clients"` HASH key

File: client/manager_redis.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+	"github.com/ory-am/fosite"
+	"github.com/ory-am/hydra/pkg"
+	"github.com/pborman/uuid"
+	"github.com/pkg/errors"
+	"gopkg.in/redis.v5"
+)
+
+type RedisManager struct {
+	DB     *redis.Client
+	Hasher fosite.Hasher
```
Comment (dim) [inline client/manager_redis.go]: I think it would be nice to support a `keyPrefix` too

File: client/manager_redis.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+const redisClientTemplate = "hydra:client:%s"
+
+func redisClientKey(id string) string {
+	return fmt.Sprintf(redisClientTemplate, id)
+}
+
+func (m *RedisManager) GetConcreteClient(id string) (*Client, error) {
+	var d Client
```
Comment (dim) [inline client/manager_redis.go]: Declare vars where you need them, it's not C

File: client/manager_redis.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+	return &d, nil
+}
+
+func (m *RedisManager) GetClient(id string) (fosite.Client, error) {
+	return m.GetConcreteClient(id)
+}
+
+func (m *RedisManager) UpdateClient(c *Client) error {
+	var s []byte
```
Comment (dim) [inline client/manager_redis.go]: Same, no need to declare `s` here.

File: client/manager_redis.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+		if err != nil {
+			return errors.Wrap(err, "")
+		}
+		c.Secret = string(h)
+	}
+	if err := mergo.Merge(c, o); err != nil {
+		return errors.Wrap(err, "")
+	}
+
+	if s, err = json.Marshal(c); err != nil {
```
Comment (dim) [inline client/manager_redis.go]:
``` go
s, err := json.Marshal(c)
if err != nil {
  ...
}
```

File: client/manager_redis.go [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+	if err := m.Hasher.Compare(c.GetHashedSecret(), secret); err != nil {
+		return nil, errors.Wrap(err, "")
+	}
+
+	return c, nil
+}
+
+func (m *RedisManager) CreateClient(c *Client) error {
+	var s []byte
```
Comment (dim) [inline client/manager_redis.go]: And again ...

File: client/manager_redis.go [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+	if err != nil {
+		return errors.Wrap(err, "")
+	}
+	c.Secret = string(hash)
+
+	if s, err = json.Marshal(c); err != nil {
+		return errors.Wrap(err, "")
+	}
+
+	if err := m.DB.Set(redisClientKey(c.ID), s, 0).Err(); err != nil {
```
Comment (dim) [inline client/manager_redis.go]: As this is `Create`, should we use SetNX (or HSetNX)?

File: client/manager_redis.go [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+	return nil
+}
+
+func (m *RedisManager) GetClients() (map[string]Client, error) {
+	clients := make(map[string]Client)
+
+	iter := m.DB.Scan(0, redisClientKey("*"), 0).Iterator()
+	for iter.Next() {
+		var d Client
```
Comment (dim) [inline client/manager_redis.go]: And again ...

File: jwk/manager_redis.go [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+	if _, err := pipe.Exec(); err != nil {
+		return errors.Wrap(err, "")
+	}
+
+	return nil
+}
+
+func (m *RedisManager) getKeySet(set string) (*jose.JsonWebKeySet, error) {
+	keySet := new(jose.JsonWebKeySet)
```
Comment (dim) [inline jwk/manager_redis.go]: `var keySet *jose.JsonWebKeySet`

File: jwk/manager_redis.go [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+
+	iter := m.DB.HScan(redisJWKKey(set), 0, "", 0).Iterator()
+	for iter.Next() {
+		var key jose.JsonWebKey
```
Comment (dim) [inline jwk/manager_redis.go]: And again ...

File: jwk/manager_redis.go [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+	}
+
+	if err := iter.Err(); err != nil {
+		return nil, err
+	}
+
+	return keySet, nil
+}
+
+func (m *RedisManager) GetKey(set, kid string) (*jose.JsonWebKeySet, error) {
```
Comment (dim) [inline jwk/manager_redis.go:95]: So this seems strange, you do have a HASH, so you could simply `HGET(set, kid)`, no?

File: jwk/manager_test.go [hunk 12]
```diff
@@ -192,6 +194,28 @@ func connectToMySQL() {
 	containers = append(containers, c)
 }
 
+func connectToRedis() {
+	var db *redis.Client
+	c, err := dockertest.ConnectToRedis(15, time.Second, func(url string) bool {
+		db = redis.NewClient(&redis.Options{
+			Addr:     url,
+			Password: "",
```
Comment (dim) [inline jwk/manager_test.go]: Please omit zero-value vars, e.g. `Password: ""` or `DB: 0`

File: jwk/manager_test.go [hunk 13]
```diff
@@ -192,6 +194,28 @@ func connectToMySQL() {
...
+			DB:       0,
+		})
+		_, err := db.Ping().Result()
```
Comment (dim) [inline jwk/manager_test.go]: `return db.Ping().Err()`

File: jwk/manager_test.go [hunk 14]
```diff
@@ -192,6 +194,28 @@ func connectToMySQL() {
...
+
+		return err == nil
+	})
+
+	if err != nil {
+		log.Fatalf("Could not connect to database: %s", err)
+	}
+
+	containers = append(containers, c)
+	s := &RedisManager{DB: db, Cipher: &AEAD{Key: encryptionKey}}
```
Comment (dim) [inline jwk/manager_test.go]: Avoid single-use vars, unless they make things easier to read, in this case:

```
managers["redis"] = &RedisManager{
  DB: db, 
  Cipher: &AEAD{Key: encryptionKey},
}
```

File: oauth2/fosite_store_redis.go [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+
+const (
+	redisOpenID  = "oidc"
+	redisAccess  = "access"
+	redisRefresh = "refresh"
+	redisCode    = "code"
+)
+
+func (s *FositeRedisStore) getKey(key string, proto fosite.Session) (*fosite.Request, error) {
+	sch := new(RedisSchema)
```
Comment (dim) [inline oauth2/fosite_store_redis.go]: And again, delacre before you actually need it as `var schema *RedisSchema`

File: oauth2/fosite_store_redis.go [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+		return nil, err
+	}
+
+	if proto != nil {
+		if err := json.Unmarshal(sch.Session, proto); err != nil {
+			return nil, errors.Wrap(err, "")
+		}
+	}
+
+	d := new(fosite.Request)
```
Comment (dim) [inline oauth2/fosite_store_redis.go]: Do

```
return &fosite.Request{
  ID: schema.ID,
  ...
}, nil
```

File: oauth2/fosite_store_redis.go [hunk 17]
```diff
@@ -0,0 +1,242 @@
...
+	RequestedAt   time.Time        `json:"requestedAt"`
+	Client        *client.Client   `json:"client"`
+	Scopes        fosite.Arguments `json:"scopes"`
+	GrantedScopes fosite.Arguments `json:"grantedScopes""`
+	Form          url.Values       `json:"form"`
+	Session       json.RawMessage  `json:"session"`
+}
+
+func redisKey(field, suffix string) string {
+	return fmt.Sprintf("hydra:oauth2:%s:%s", field, suffix)
```
Comment (dim) [inline oauth2/fosite_store_redis.go]: Can we use HASHes e.g. `hydra:oauth2:oidc` and `hydra:oauth2:code`?

File: oauth2/fosite_store_redis.go [hunk 18]
```diff
@@ -0,0 +1,242 @@
...
+	d.Client = sch.Client
+	d.Scopes = sch.Scopes
+	d.GrantedScopes = sch.GrantedScopes
+	d.Form = sch.Form
+	d.Session = proto
+
+	return d, nil
+}
+
+func (s *FositeRedisStore) setKey(key string, requester fosite.Requester) error {
```
Comment (dim) [inline oauth2/fosite_store_redis.go]: rename to `hSet`

File: oauth2/fosite_store_redis.go [hunk 19]
```diff
@@ -0,0 +1,242 @@
...
+}
+
+const (
+	redisOpenID  = "oidc"
+	redisAccess  = "access"
+	redisRefresh = "refresh"
+	redisCode    = "code"
+)
+
+func (s *FositeRedisStore) getKey(key string, proto fosite.Session) (*fosite.Request, error) {
```
Comment (dim) [inline oauth2/fosite_store_redis.go]: rename to `hGet`

File: oauth2/fosite_store_redis.go [hunk 20]
```diff
@@ -0,0 +1,242 @@
...
+	"golang.org/x/net/context"
+	"gopkg.in/redis.v5"
+)
+
+type FositeRedisStore struct {
+	client.Manager
+	DB *redis.Client
+}
+
+type RedisSchema struct {
```
Comment (dim) [inline oauth2/fosite_store_redis.go]: Does this need to be public?

File: oauth2/fosite_store_redis.go [hunk 21]
```diff
@@ -0,0 +1,242 @@
...
+	return s.DB.Del(redisKey(redisCode, code)).Err()
+}
+
+func (s *FositeRedisStore) CreateAccessTokenSession(_ context.Context, signature string, req fosite.Requester) error {
+	err := s.setKey(redisKey(redisAccess, signature), req)
+	if err != nil {
+		return errors.Wrap(err, "")
+	}
+
+	return s.DB.SAdd(redisKey(redisAccess, req.GetID()), signature).Err()
```
Comment (dim) [inline oauth2/fosite_store_redis.go]: I think this keys should have an additional level of namespace, e.g. an `hydra:oauth2:access` HASH to store all values and an index named `hydra:oauth2:access:members:ID`

File: oauth2/fosite_store_redis.go [hunk 22]
```diff
@@ -0,0 +1,242 @@
...
+		return nil, errors.Wrap(fosite.ErrNotFound, "")
+	} else if err != nil {
+		return nil, errors.Wrap(err, "")
+	}
+
+	return session, nil
+}
+
+func (s *FositeRedisStore) DeleteAccessTokenSession(_ context.Context, signature string) error {
+	accessTokenSession, err := s.getKey(redisKey(redisAccess, signature), nil)
```
Comment (dim) [inline oauth2/fosite_store_redis.go]: `sess, err :=` is verbose enough :)

File: oauth2/fosite_store_redis.go [hunk 23]
```diff
@@ -0,0 +1,242 @@
...
+	if err == redis.Nil {
+		return nil
+	} else if err != nil {
+		return errors.Wrap(err, "")
+	}
+
+	if err := s.DB.Del(redisKey(redisAccess, signature)).Err(); err != nil {
```
Comment (dim) [inline oauth2/fosite_store_redis.go]: Please do the Del and the SRem in a pipeline

File: oauth2/fosite_store_redis.go [hunk 24]
```diff
@@ -0,0 +1,242 @@
...
+
+func (s *FositeRedisStore) DeleteRefreshTokenSession(_ context.Context, signature string) error {
+	refreshTokenSession, err := s.getKey(redisKey(redisRefresh, signature), nil)
+	if err == redis.Nil {
+		return nil
+	} else if err != nil {
+		return errors.Wrap(err, "")
+	}
+
+	if err := s.DB.Del(redisKey(redisRefresh, signature)).Err(); err != nil {
```
Comment (dim) [inline oauth2/fosite_store_redis.go]: Please do the Del and the SRem in a pipeline

File: oauth2/fosite_store_redis.go [hunk 25]
```diff
@@ -0,0 +1,242 @@
...
+	} else if err := s.CreateRefreshTokenSession(ctx, refreshSignature, req); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *FositeRedisStore) RevokeRefreshToken(ctx context.Context, id string) error {
+	iter := s.DB.SScan(redisKey(redisRefresh, id), 0, "", 0).Iterator()
+	for iter.Next() {
```
Comment (dim) [inline oauth2/fosite_store_redis.go:255]: start a `pipe := s.DB.Pipeline()` here

File: oauth2/fosite_store_redis.go [hunk 26]
```diff
@@ -0,0 +1,242 @@
...
+		sig := iter.Val()
+		if err := s.DeleteRefreshTokenSession(ctx, sig); err != nil {
```
Comment (dim) [inline oauth2/fosite_store_redis.go]: You don't need to do that, you can just `pipe.HDel(..., sig)` here

File: oauth2/fosite_store_redis.go [hunk 27]
```diff
@@ -0,0 +1,242 @@
...
+			return err
+		}
+	}
+
+	return iter.Err()
+}
```
Comment (dim) [inline oauth2/fosite_store_redis.go:265]: And `pipe.Del(...)` to remove the SET key in the end

File: oauth2/fosite_store_redis.go [hunk 28]
```diff
@@ -0,0 +1,242 @@
...
+
+func (s *FositeRedisStore) RevokeAccessToken(ctx context.Context, id string) error {
+	iter := s.DB.SScan(redisKey(redisAccess, id), 0, "", 0).Iterator()
```
Comment (dim) [inline oauth2/fosite_store_redis.go]: Same as `RevokeRefreshToken`

File: oauth2/fosite_store_redis.go (hunk 23 above)
Comment (115100) [reply oauth2/fosite_store_redis.go]: Should I use a pipe for hSet/SAdd during creation too?

**Comment by dim** [comment]
Please

File: oauth2/fosite_store_redis.go [hunk 29]
```diff
@@ -0,0 +1,273 @@
...
+	} else if err != nil {
+		return errors.Wrap(err, "")
+	}
+
+	pipe := s.DB.Pipeline()
+	defer pipe.Close()
+
+	pipe.HDel(s.redisKey(redisRefresh), signature)
+	pipe.SRem(s.redisKey(redisRefresh, sess.GetID()), signature)
+	if _, err := pipe.Exec(); err != nil {
```
Comment (dim) [inline oauth2/fosite_store_redis.go]:
```go
_, err := pipe.Exec()
return err
```

File: oauth2/fosite_store_redis.go [hunk 30]
```diff
@@ -0,0 +1,273 @@
...
+
+const (
+	redisHydraKey = "hydra:oauth2"
+	redisOpenID   = "oidc"
+	redisAccess   = "access"
+	redisRefresh  = "refresh"
+	redisCode     = "code"
+)
+
+func createTokenSession(pipe *redis.Pipeline, req fosite.Requester, key, setKey, signature string) error {
```
Comment (dim) [inline oauth2/fosite_store_redis.go]: rename to `redisCreateTokenSession`

**Review by dim — APPROVED** [review]
Minor, otherwise :+1:

**Pull request closed without merging**